src/extension/type_template.rs

```diff
@@ -0,0 +1,353 @@
+//! Type schemes that can be turned into [FunctionType]s by applying
+//! them to [TypeArg]s
+
+use std::collections::HashMap;
+
+use smol_str::SmolStr;
+
+use super::{ExtensionId, ExtensionSet, SignatureError, TypeDefBound, TypeParametrised};
+use crate::types::type_param::{check_type_arg, TypeArgError};
+use crate::types::TypeBound;
+use crate::{ops::AliasDecl, Extension};
+
+use crate::types::{
+    type_param::{CustomTypeArg, TypeArg, TypeParam},
+    CustomType, FunctionType, Type, TypeRow,
+};
+
+/// deBruijn-indexed Variables
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+struct VariableRef(usize);
+
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+enum ExtensionSetTemplate {
+    // TODO possibly allow union of ExtensionSet with any number of vars?
+    Concrete(ExtensionSet),
+    TypeVar(VariableRef),
+}
+
+/// TypeEnum with Prim inlined (as we need our own version of FunctionType)
+/// and including variables
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+enum TypeTemplate {
+    Extension(CustomTypeTemplate),
+    Alias(AliasDecl),
+    Function(FunctionTypeTemplate),
+    Tuple(Vec<TypeTemplate>),
+    Sum(Vec<TypeTemplate>),
+    TypeVar(VariableRef),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+struct FunctionTypeTemplate {
+    input: Vec<TypeTemplate>,
+    output: Vec<TypeTemplate>,
+    extension_reqs: ExtensionSetTemplate,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+struct CustomTypeTemplate {
+    extension: ExtensionId,
+    id: SmolStr,
+    args: Vec<TypeArgTemplate>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[non_exhaustive]
+enum TypeArgTemplate {
+    Type(TypeTemplate),
+    USize(u64),
+    Opaque(CustomTypeArg),
+    Sequence(Vec<TypeArgTemplate>),
+    Extensions(ExtensionSet),
+    TypeVar(VariableRef),
+}
+
+/// Representation of the type scheme for an Op with declared TypeParams.
+/// This should be the result of YAML parsing.
+pub struct SignatureTemplate(pub Vec<TypeParam>, FunctionTypeTemplate);
```

Comment:
anything stopping this being 
```suggestion
pub struct SignatureTemplate(pub Vec<TypeParam>, TypeTemplate);
```
even if we don't provide syntax for all types and keep it to function types, seems the logic is generally applicable